util/clusterauth/clusterauth.go

```diff
@@ -215,39 +216,103 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
 	return GetServiceAccountBearerToken(clientset, ns, ArgoCDManagerServiceAccount)
 }
 
-// GetServiceAccountBearerToken will attempt to get the provided service account until it
-// exists, iterate the secrets associated with it looking for one of type
-// kubernetes.io/service-account-token, and return it's token if found.
+// GetServiceAccountBearerToken determine if a ServiceAccount has a
+// bearer token sercet to use or if one should be created. It then
+// waits for the secret to have a bearer token if a secret needs to
+// be created and returns the token.
 func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
+	secretName, err := createServiceAccountTokenSecret(clientset, sa, ns)
+	if err != nil {
+		return "", err
+	}
+
 	var serviceAccount *corev1.ServiceAccount
 	var secret *corev1.Secret
-	var err error
 	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("Failed to retrieve secret %q: %v", secretName, err)
+		}
+
+		_, ok := secret.Data["token"]
+		if !ok {
+			return false, fmt.Errorf("Secret %q for service account %q did not have a token", secret.Name, serviceAccount)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// createServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func createServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.Poll(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		if getErr != nil {
+			return false, getErr
+		}
+		if len(serviceAccount.Secrets) == 0 {
+			return false, nil
 		}
-		// Scan all secrets looking for one of the correct type:
-		for _, oRef := range serviceAccount.Secrets {
-			var getErr error
-			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
+		return true, nil
+	})
+	if errors.IsNotFound(err) {
+		return "", err
+	}
+
+	if len(serviceAccount.Secrets) != 0 {
+		for _, s := range serviceAccount.Secrets {
+			existingSecret, err := clientset.CoreV1().Secrets(ns).Get(context.Background(), s.Name, metav1.GetOptions{})
```

Comment:
Seems like we should set a timeout.

Reply:
Any timeout amount make sense to you?

Reply:
How about an outside-the-loop context with a `30s` timeout and an inside-the-loop context with a 10s timeout?

If for some absurd reason there are 100 secrets, the outer context will keep us from waiting forever.

If there's high latency, inner context will make sure we give a nice amount of time to get the secret.

If there's high latency _and_ 3+ secrets _and_ the target secret isn't in the first 3... well that's just a bummer.